Feat/pull mplex

dryajov opened the pull request:
UPDATE: This is now ready for review.

TODO:

- [x] Performance
  - [ ] pool channel objects (should come next, but not critical for first release)

Bellow are the accompanying PRs:

- [ ] https://github.com/libp2p/js-libp2p-mplex/pull/76
- [ ] https://github.com/libp2p/js-libp2p-switch/pull/254
- [ ] https://github.com/libp2p/js-libp2p/pull/192
- [ ] https://github.com/ipfs/js-ipfs/pull/1346

---

## Perf:

I was able to shave off ~15 mins of the new implementations running the mega stress tests, the perf is now about the same (or a little better ;) ) than the stream based implementation.

<img width="930" alt="screen shot 2018-04-27 at 7 13 30 pm" src="https://user-images.githubusercontent.com/1094341/39390405-53fdcaae-4a51-11e8-8aa2-3ad6741b843a.png">


[Archive.zip](https://github.com/libp2p/pull-mplex/files/1957136/Archive.zip)

The attached zip contains a heap snapshot and a perf log taken with `node --prof --nologfile_per_isolate --logfile=xxxxxx.log --log-timer-events /private/var/folders/_r/6c6jf6m10kb3v9kt45qspw4w0000gn/T/v8profilerProxy.js`. The one thing that still irks me a bit is the excessive (compared to the prev implementation) GC time. I think we can improve that a lot by pooling the `Channel` objects and reusing them, rather than creating new ones every time.

Here is a screenshot from the perf log graph - GC is seems to take ~13% of the total time. (These graphs where generated using the attached logs, and webstorm perf and heap analyzers).

<img width="1387" alt="screen shot 2018-04-27 at 7 28 07 pm" src="https://user-images.githubusercontent.com/1094341/39390414-7d889bc4-4a51-11e8-8d42-a03a385c3600.png">

<img width="1387" alt="screen shot 2018-04-27 at 7 27 54 pm" src="https://user-images.githubusercontent.com/1094341/39390415-7da28c82-4a51-11e8-98e0-6cd6f0187342.png">


Keep having issues with this PR, use this instead of #1, #2, #3.

dryajov committed:
feat: initial commit

dryajov committed:
feat: correct channel closing

dryajov committed:
feat: getting interface-stream-muxer tests to pass

dryajov committed:
feat: getting interface-stream-muxer tests to pass

dryajov committed:
feat: update README.md

dryajov committed:
feat: improve performance

dryajov committed:
feat: correct encoding/decoding

dryajov committed:
feat: correct encoding and channel closing

dryajov committed:
feat: correct channel and plex closing

dryajov committed:
feat: interop with old libp2p-mplex

dryajov committed:
fix: correctly read chunked data

dryajov committed:
feat: interep with old stream based mplex

dryajov committed:
feat: avoid negative sizes

dryajov committed:
feat: dont close channel too early

dryajov committed:
feat: validation and coverage

dryajov committed:
fix: correct channel id sequence

dryajov committed:
feat: use bl to avoid buffer concat/copy

dryajov committed:
chore: renaming back to pull-mplex

dryajov committed:
feat: initial commit

dryajov committed:
chore: rename old to stream-mplex-interop

dryajov committed:
chore: adding misc files

dryajov committed:
chore: license

dryajov committed:
feat: use two lists to avoid clashes

dryajov committed:
style: formatting

dryajov committed:
feat: remove unused deps

dryajov committed:
feat: performance

dryajov committed:
chore: update repo link

dryajov committed:
fix: dont use bl

dryajov committed:
perf: optimize

dryajov committed:
style: fix formating

dryajov committed:
perf: improoving performance

dryajov committed:
style: remove unused variable

dryajov committed:
fix: msg length encoding

dryajov committed:
fix: dont blow up the stack

dryajov committed:
fix: use arrow function

dryajov committed:
fix: remove initiator id logic

dryajov committed:
fix: start channel from 0

dryajov committed:
chore: adding commitizen

dryajov committed:
chore: remove unused file

dryajov committed:
feat: send chan name

dryajov committed:
fix: send chan name when creating stream

dryajov committed:
fix: remove nyc_output file

dryajov committed:
chore: move benchmarks and inteop to libp2p-mplex

dryajov commented:
@libp2p/javascript-team is this still of any interest to the libp2p team, if not can I publish this as a standalone module?

jacobheun commented:
@dryajov it is! We were talking about this last week. I'd like to try and get this finished up in the next few weeks in relation to https://github.com/libp2p/js-libp2p-mplex/issues/89

dryajov commented:
Awesome! I'm here to assist with whatever I can.

jacobheun committed:
chore: remove non jenkins ci

jacobheun commented:
@dryajov just a heads up, I'm working on finalizing this this week. I hit some errors (oddly when updating just aegir) that I'm working through.

I also think it might be beneficial to have this implement the `interface-stream-muxer` interface. We could close libp2p/js-libp2p-mplex#76 and just implement this as a standalone muxer, then swap out it's usage in the libp2p and js-ipfs bundles. Do you think there is value in keeping the existing approach vs doing that?

jacobheun committed:
chore: update dependencies

jacobheun committed:
fix: listen to the channel error

need to figure out the best way to handle the error and if we should be emitting the errors

jacobheun commented on src/index.js:
```diff
@@ -186,6 +186,12 @@ class Mplex extends EE {
+      // is handling errors. In node 10+ if we dont catch this
+      // error, it will be thrown. Perhaps channel shouldnt be
+      // emitting errors?
+    })
```
This was needed to fix an error with the test suite, because there was no error listener on channel. I need to do some more testing, but I don't think the channel needs to be emitting errors. It appears to be handling everything internally and reseting the stream properly, so I think it's probably redundant anyway.

mcollina replied:
I think the Channel should be emitting errors, but those needs to be catched here and logged. A lot of channel erroring might be a smell than it's better to tear down the connection?

jacobheun committed:
refactor: spread op instead of lodash defaults

jacobheun committed:
refactor: avoid toString when debug not enabled

jacobheun committed:
docs: add more comments

refactor: update some code for clarity

jacobheun committed:
docs: add more js docs

refactor: clean up code for clarity

jacobheun commented:
Dependencies are updated and I went ahead and added jsdocs and some comments across the project as I was reviewing things. Aside from the channel object pooling, that was already mentioned, I think this looks good.

My thoughts on moving this forward are:
1. Merge this
2. Do a subsequent PR to add in the interface-stream-muxer
3. Do a release candidate
4. Test against switch, libp2p, js-ipfs, and ipfs/interop, replacing libp2p-mplex with this.
5. Get this released up the chain, replacing libp2p-mplex 🎉 
6. Do a followup release with channel object pooling

The reason I'd like to have this as a standalone from libp2p-mplex, is that in the future, we might indeed change that over to stream3/async iterators. However, that's a lot of work and is going to happen soon. I'd like to keep the benchmarks added in [libp2p/js-libp2p-mplex#76](https://github.com/libp2p/js-libp2p-mplex/pull/76), so that we can test the pull-stream implementation against stream3 before doing a full move in that direction.

Can I get some feedback on this approach before moving that direction? @dryajov @alanshaw @vasco-santos @mkg20001 @daviddias ❤️

dryajov commented:
👍 on overall plan, implementing interface-stream-muxer also makes sense.

mcollina commented on src/channel.js:
```diff
@@ -0,0 +1,244 @@
+          this.sendMsg(data)
+          next()
+        })
+      })
```
The above creates a lot of closures. I'd recommend to refactor this logic into a top level or instance function to ease V8 optimizer job.

mcollina commented on src/channel.js:
```diff
@@ -0,0 +1,244 @@
+   * Closes the channel with the given error
+   * @param {Error} err Default: `'channel reset!'`
+   */
+  reset (err) {
```
This is often called `destroy()` in Node lingo. Should this have a callback?

mcollina commented on src/channel.js:
```diff
@@ -0,0 +1,244 @@
+   *
+   * @param {Buffer} data
+   */
+  sendMsg (data) {
```
Shouldn't this have a callback?

mcollina commented on src/coder.js:
```diff
@@ -0,0 +1,166 @@
+    this.queue(pool.slice(oldUsed, used)) // send header
+
+    if (PULL_LENGTH - used < 100) {
+      pool = Buffer.alloc(PULL_LENGTH)
```
I would be bold and go for `allocUnsafe`. We are writing 100% of the memory, minus bugs.

mcollina commented on src/coder.js:
```diff
@@ -0,0 +1,166 @@
+      // Reading is done for this message, start processing it
+      if (States.PARSING === state) {
+        if (accumulating) {
+          used += msg.copy(buffer, used)
```
It's better to accumulate the buffer in an array and just concat them all. All this copying will come at a cost.

mcollina commented on src/index.js:
```diff
@@ -0,0 +1,315 @@
+      // Create a new stream
+      case Types.NEW: {
+        const chan = this._newStream(id, false, true, data.toString(), this._inChannels)
+        setImmediate(() => this.emit('stream', chan, id))
```
why `setImmediate` and not `nextTick`?

mcollina commented on src/index.js:
```diff
@@ -0,0 +1,315 @@
+      }
+
+      default:
+        this.emit('error', new Error('Invalid message type'))
```
this might need to be deferred with `nextTick`.

mcollina reviewed (commented):
I think it would be good to have some quick benchmarks built with benchmark.js or fastbench to check if everything is working correctly. Also, we should check a) flamegraphs and b) bubblegraphs if there is any areas we could improve. These should also serve as a comparison between the two implementations. You might grab some of the code in https://github.com/mcollina/tentacoli/tree/master/benchmarks as an example.

mcollina commented:
(Benchmarks could also come in a separate PR).

jacobheun commented:
Thanks for the feedback @mcollina. I made a couple small changes (nextTick and error logging). I am going to merge this and work on the performance and callback items you noted in a new PR that will include benchmarks and the `interface-stream-muxer` additions.
